### ISSUE TITLE
[MNG-7511] Ensure the degreeOfConcurrency is a positive number in MavenExecutionRequest

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8950]
**[Josef Cacek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwart)** opened **[MNG-7511](https://issues.apache.org/jira/browse/MNG-7511?redirect=false)** and commented

The `degreeOfConcurrency` should always be a positive number, otherwise builds might fail.

In Hazelcast we build with threads set to `-T 0.5C` in [`.mvn/maven.config`](https://github.com/hazelcast/hazelcast/blob/v5.1.2/.mvn/maven.config).
This is fine until we run the build on a single CPU machine. It fails then with `IllegalArgumentException` as the computed `degreeOfConcurrency == 0`.

Stacktrace from Maven 3.6.3:

```
[ERROR] IllegalArgumentException
java.lang.IllegalArgumentException
    at java.util.concurrent.ThreadPoolExecutor.<init> (ThreadPoolExecutor.java:1293)
    at java.util.concurrent.ThreadPoolExecutor.<init> (ThreadPoolExecutor.java:1215)
    at java.util.concurrent.Executors.newFixedThreadPool (Executors.java:155)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.build (MultiThreadedBuilder.java:88)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

---

**Affects:** 3.8.6

**Remote Links:**
- [GitHub Pull Request #767
](https://github.com/apache/maven/pull/767)
- [GitHub Pull Request #773
](https://github.com/apache/maven/pull/773)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1), [3.8.8](https://github.com/apache/maven/milestone/109?closed=1)

**Comment by jira-importer** [comment]
**[Josef Cacek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwart)** commented

Pull request with a suggested fix is here: https://github.com/apache/maven/pull/767

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

3.6.3 or 3.8.6? Does it reproduce with 3.8.6?

**Comment by jira-importer** [comment]
**[Josef Cacek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwart)** commented

Reproduced with the 3.8.6:

```
[ERROR] IllegalArgumentException
java.lang.IllegalArgumentException
    at java.util.concurrent.ThreadPoolExecutor.<init> (ThreadPoolExecutor.java:1293)
    at java.util.concurrent.ThreadPoolExecutor.<init> (ThreadPoolExecutor.java:1215)
    at java.util.concurrent.Executors.newFixedThreadPool (Executors.java:155)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.build (MultiThreadedBuilder.java:90)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [03b1faff7eec915f8b277d307d3e3aa0d0a24fdf](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=03b1faff7eec915f8b277d307d3e3aa0d0a24fdf).
